secure/storage/src/vault.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    error::Error, permissions::Permissions, storage::Storage, value::Value,
+    vault_client::VaultClient,
+};
+
+/// VaultStorage utilizes Vault for maintaining secret data for Libra. This version currently
+/// matches the behavior of OnDiskStorage and InMemoryStorage. In the future, Vault will be able to
+/// create keys, sign messages, and handle permissions across different services.
+pub struct VaultStorage {
+    client: VaultClient,
+}
+
+impl VaultStorage {
+    pub fn new(host: String, token: String) -> Self {
+        Self {
+            client: VaultClient::new(host, token),
+        }
+    }
+
+    pub fn reset(&self) -> Result<(), Error> {
+        let secrets = self.client.list_secrets("")?;
+        for secret in secrets {
+            self.client.delete_secret(&secret)?;
+        }
+        Ok(())
+    }
+
+    fn get_secret(&self, key: &str) -> Result<Value, Error> {
+        let value = self.client.read_secret(key, key)?;
+        let v = Value::from_base64(&value).unwrap();
+        Ok(v)
+    }
+
+    fn set_secret(&self, key: &str, value: Value) -> Result<(), Error> {
+        self.client.write_secret(key, key, &value.to_base64()?)?;
+        Ok(())
+    }
+}
+
+impl Storage for VaultStorage {
+    fn available(&self) -> bool {
+        self.client.unsealed().unwrap_or(false)
+    }
+
+    fn create(&mut self, key: &str, value: Value, _permissions: &Permissions) -> Result<(), Error> {
```

Comment:
the permissions will be used to verify the client.token?

Reply:
yes, I have a few more commits not released that actually make sense of this part of the framework :).